aws_crt_common_ws/aws-crt-common/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+                stringify!($error_code),
+                ", ",
+                $error_str
+            )),
+        }
+    };
+}
+
+/// Default allocator to use for most of this crate's APIs. This is usually the one
+/// you want, unless you're testing or profiling.
+#[non_exhaustive]
+pub struct DefaultAllocator {
+    c_allocator: *const CAllocator,
+}
+
+impl DefaultAllocator {
+    pub fn new() -> DefaultAllocator {
+        // aws_default_allocator() always returns a valid address to static memory.
+        // it does not need a destructor since the allocator is static memory.
+        unsafe {
+            let alloc = DefaultAllocator {
+                c_allocator: aws_default_allocator(),
+            };
+            return alloc;
+        }
+    }
+}
+
+impl NativeResource<CAllocator> for DefaultAllocator {
+    fn get_native_resource(&self) -> *const CAllocator {
+        self.c_allocator
+    }
+}
+
+/// This is a good debugging allocator that will provide diagnositics on memory errors and leaks.
+///
+/// This is most useful for tests.
+#[non_exhaustive]
+pub struct TracingAllocator {
+    c_allocator: *const CAllocator,
+}
+
+impl TracingAllocator {
+    pub fn new(level: TraceLevel, frames_per_stack: usize) -> TracingAllocator {
+        // aws_default_allocator() always returns a valid address and it's static memory so it
+        // doesn't need to be cleaned up. However, aws_mem_tracer_new() allocates under the hood,
+        // and this memory will be reclaimed in the destructor. If the destructor doesn't run,
+        // the worst that will happen is a memory leak.
+        unsafe {
+            let default_allocator = aws_default_allocator();
+            let alloc = TracingAllocator {
+                c_allocator: aws_mem_tracer_new(
+                    default_allocator,
+                    null_mut(),
+                    level,
+                    frames_per_stack,
+                ),
+            };
+            return alloc;
+        }
+    }
+
+    /// returns the count of currently allocated bytes.
+    pub fn allocated_bytes(&self) -> usize {
+        // If we have an instance of this class, c_allocator is assumed valid.
+        // simply returns the count of currently allocated bytes.
+        unsafe {
+            return aws_mem_tracer_bytes(self.c_allocator);
+        }
+    }
+}
+
+impl NativeResource<CAllocator> for TracingAllocator {
+    fn get_native_resource(&self) -> *const CAllocator {
+        self.c_allocator
+    }
+}
+
+impl Drop for TracingAllocator {
+    fn drop(&mut self) {
+        // aws_mem_tracer_destroy() returns the wrapped allocator, but since in new()
+        // it's specified as the default_allocator, it doesn't need to be cleaned up.
+        unsafe {
+            let _ = aws_mem_tracer_destroy(self.c_allocator);
+        }
+    }
+}
+
+/// String type which wraps the internal aws_string structure. This type is immutable.
+#[non_exhaustive]
+pub struct AwsString {
```

Comment:
You will probably want to add `Deref` impls with `str` as the target. You will get a bunch of methods for "free". See: https://doc.rust-lang.org/stable/std/string/struct.String.html#deref-methods

You will get `as_ptr()` as well, which you implemented ass `bytes()`.

Reply:
yeah, this con't compile, and the docs got nothing. I'll come back to it at a later date.